### PR TITLE
Enable errorlint via golangci-lint

### PR DESCRIPTION
**Pull request by rosstimothy**
Adds the new linter and configures it to prevent direct error comparisions and type assertions in favor of errors.Is and errors.As. All current violations caught by the linter have been updated accordingly.

Depends on https://github.com/gravitational/teleport.e/pull/3348

### CODE REVIEW

File: .golangci.yml [hunk 1]
```diff
@@ -58,6 +59,10 @@ linters-settings:
             desc: 'experimental project, not to be confused with official Go packages'
           - pkg: golang.org/x/exp/slices
             desc: 'use "slices" instead'
+  errorlint:
+    comparison: true
+    asserts: true
+    errorf: false
```
Comment (codingllama) [inline .golangci.yml:65]: What happens if we turn this on? Too many occurrences?

File: examples/dynamoathenamigration/migration.go [hunk 2]
```diff
@@ -723,7 +723,7 @@ func mergeFiles(files []*os.File, outputFile *os.File) error {
 		}
 
 		nextLine, err := readLine(min.Dec)
-		if trace.Unwrap(err) == io.EOF {
+		if errors.Is(trace.Unwrap(err), io.EOF) {
```
Comment (codingllama) [inline examples/dynamoathenamigration/migration.go:726]: I think we don't need to trace.Unwrap before errors.Is or errors.As, they should do the unwrapping by themselves. That said I think it's fine to keep the lines close to the original too.

File: integrations/access/opsgenie/app.go [hunk 3]
```diff
@@ -236,7 +236,7 @@ func (a *App) onPendingRequest(ctx context.Context, req types.AccessRequest) err
 		return nil
 	}
 	// Don't show the error if the annotation is just missing.
-	if trace.Unwrap(notifyErr) == errMissingAnnotation {
+	if errors.Is(notifyErr, trace.Unwrap(notifyErr)) {
```
Comment (codingllama) [inline integrations/access/opsgenie/app.go:239]:
```suggestion
	if errors.Is(trace.Unwrap(notifyErr), errMissingAnnotation) {
```

File: lib/multiplexer/multiplexer.go [hunk 4]
```diff
@@ -321,7 +321,7 @@ func (m *Mux) detectAndForward(conn net.Conn) {
 
 	connWrapper, err := m.detect(conn)
 	if err != nil {
-		if trace.Unwrap(err) != io.EOF {
+		if !errors.Is(trace.Wrap(err), io.EOF) {
```
Comment (codingllama) [inline lib/multiplexer/multiplexer.go:324]:
```suggestion
		if !errors.Is(trace.Unwrap(err), io.EOF) {
```

File: lib/srv/db/common/errors.go [hunk 5]
```diff
@@ -47,29 +47,39 @@ func ConvertError(err error) error {
...
-	case *pgconn.PgError:
-		return convertPostgresError(e)
-	case *mysql.MyError:
-		return convertMySQLError(e)
+
+	var googleAPIErr *googleapi.Error
+	var awsRequestFailureErr awserr.RequestFailure
+	var azResponseErr *azcore.ResponseError
+	var pgError *pgconn.PgError
+	var myError *mysql.MyError
```
Comment (codingllama) [inline lib/srv/db/common/errors.go:71]: This is a bit sad, as each errors.As invocation is a separate traversal of the error chain. Hopefully the chains are small enough that this won't make much of a difference.

File: api/utils/grpc/interceptors/errors_test.go [hunk 6]
```diff
@@ -99,15 +101,16 @@ func TestGRPCErrorWrapping(t *testing.T) {
 		// io.EOF error (meaning the server errored and closed the stream).
 		// In either case, it is still safe to recv from the stream and check for
 		// the already exists error.
-		if sendErr != nil && sendErr != io.EOF /* == error comparison on purpose! */ {
+		//nolint:errorlint //comparison error comparison on purpose!
```
Comment (codingllama) [inline api/utils/grpc/interceptors/errors_test.go:104]:
```suggestion
		//nolint:errorlint // == error comparison on purpose!
```

**Review by codingllama — COMMENTED** [review]
Thanks for tackling this!

File: .golangci.yml (hunk 1 above)
Comment (codingllama) [reply .golangci.yml:65]: Sounds good. If you prefer doing a separate PR that would be fine too.

Comment (codingllama) [reply .golangci.yml]: Nice, this should keep me honest.  :)

File: api/utils/retryutils/retry.go [hunk 7]
```diff
@@ -185,7 +186,8 @@ func (r *Linear) For(ctx context.Context, retryFn func() error) error {
 		if err == nil {
 			return nil
 		}
-		if _, ok := trace.Unwrap(err).(*permanentRetryError); ok {
+		var permanentRetryError *permanentRetryError
+		if errors.As(trace.Unwrap(err), &permanentRetryError) {
```
Comment (strideynet) [inline api/utils/retryutils/retry.go:190]: Does this need the Unwrap here if we are using errors.As ?

**Review by strideynet — APPROVED** [review]

File: api/utils/retryutils/retry.go (hunk 7 above)
Comment (codingllama) [reply api/utils/retryutils/retry.go:190]: Similar thread at https://github.com/gravitational/teleport/pull/37669#discussion_r1475118680.

We don't need the unwraps, but I think Tim was avoiding changing too much.

File: e
Comment (codingllama) [inline e:1]: Friendly reminder to update this to a "proper" commit when possible.

**Review by codingllama — APPROVED** [review]

File: api/utils/retryutils/retry.go (hunk 7 above)
Comment (strideynet) [reply api/utils/retryutils/retry.go:190]: Ah - fair enough - I missed that comment since it was resolved 🙈

**Pull request merged**